File: LeetCode/419-battleships-in-a-board/ex_419.cpp
class Solution {
public:
    int countBattleships(vector<vector<char>>& board) {
        int count = 0;
        for (int i = 0; i != board.size(); ++i) {
            for (int j = 0; j != board[0].size(); ++j) {
                if (shouldCount(board, i, j))
                    ++count;
            }
        }
        return count;
    }
    
    bool hasBattleship(vector<vector<char>>& board, int x, int y) {
        if (x < 0 || y < 0)     return false;
        return board[x][y] == 'X';
    }
    
    bool shouldCount(vector<vector<char>>& board, int x, int y) {
        return hasBattleship(board, x, y) && !hasBattleship(board, x, y - 1) && !hasBattleship(board, x - 1, y);
    }
};

File: LeetCode/419-battleships-in-a-board/ex_419.py
class Solution(object):
    def countBattleships(self, board):
        """
        :type board: List[List[str]]
        :rtype: int
        """
        def has_battleship(x, y):
            if x < 0 or y < 0:  return False
            return board[x][y] == 'X'
        
        def should_count(x, y):
            return has_battleship(x, y) and not has_battleship(x, y - 1) and not has_battleship(x - 1, y)
        
        count = 0
        for x in xrange(len(board)):
            for y in xrange(len(board[0])):
                if should_count(x, y):  count += 1
        return count
        